ruby/ql/lib/codeql/ruby/frameworks/ActionDispatch.qll

```diff
@@ -0,0 +1,855 @@
+
+    override string getAction() { result = action }
+
+    override string getHTTPMethod() { result = httpMethod }
+  }
+
+  /**
+   * A route generated by a call to `resource`.
+   * This is like a `resources` route, but creates routes for a singular resource.
+   * This means there's no index route, no id parameter, and the resource name is expected to be singular.
+   * It will still be routed to a pluralised controller name.
+   * ```ruby
+   * resource :account
+   * ```
+   */
+  private class SingularResourceRoute extends Route, TResourceRoute {
+    RouteBlock parent;
+    string resource;
+    string action;
+    string httpMethod;
+    string pathComponent;
+
+    SingularResourceRoute() {
+      this = TResourceRoute(parent, method, action) and
+      resource = method.getArgument(0).(StringlikeLiteral).getValueText() and
+      isDefaultSingularResourceRoute(resource, httpMethod, pathComponent, action)
+    }
+
+    override string getAPrimaryQlClass() { result = "SingularResourceRoute" }
+
+    override RouteBlock getParentBlock() { result = parent }
+
+    override string getLastPathComponent() { result = pathComponent }
+
+    override string getLastControllerComponent() { result = method.getArgument(0).getValueText() }
+
+    override string getAction() { result = action }
+
+    override string getHTTPMethod() { result = httpMethod }
+  }
+
+  /**
+   * A route generated by a call to `match`.
+   * This is a lower level primitive that powers `get`, `post` etc.
+   * The first argument can be a path or a (path, controller-action) pair.
+   * The controller, action and HTTP method can be specified with the
+   * `controller:`, `action:` and `via:` keyword arguments, respectively.
+   * ```ruby
+   * match 'photos/:id' => 'photos#show', via: :get
+   * match 'photos/:id', to: 'photos#show', via: :get
+   * match 'photos/:id', controller: 'photos', action: 'show', via: :get
+   * ```
+   */
+  private class MatchRoute extends Route, TMatchRoute {
+    private RouteBlock parent;
+
+    MatchRoute() { this = TMatchRoute(parent, method) }
+
+    override string getAPrimaryQlClass() { result = "MatchRoute" }
+
+    override RouteBlock getParentBlock() { result = parent }
+
+    override string getLastPathComponent() {
+      result = method.getArgument(0).(StringlikeLiteral).getValueText() or
+      result = method.getArgument(0).(Pair).getKey().getValueText()
+    }
+
+    override string getLastControllerComponent() {
+      result = extractController(method.getKeywordArgument("to").getValueText()) or
+      result = method.getKeywordArgument("controller").getValueText() or
+      result = extractController(method.getArgument(0).(Pair).getValue().getValueText())
+    }
+
+    override string getHTTPMethod() {
+      result = method.getKeywordArgument("via").(StringlikeLiteral).getValueText() or
+      result = method.getKeywordArgument("via").(ArrayLiteral).getElement(_).getValueText()
```

Comment:
Wonder if this should have a special case for `:all` where the `result = ["get", "post", "put", "patch", "delete"]` (https://guides.rubyonrails.org/routing.html#http-verb-constraints)?